### ISSUE TITLE
Fix S3060 FP: Switch on this with new pattern matching

### ISSUE THREAD

**Issue by prinz7** [SonarSource/sonar-dotnet#7093]
### Description

Using a return switch statement with pattern matching on this fires a false positive of rule S3060.
It is a false positive, because there is no type check on a subclass. Only values from the same class are checked. See example below:

### Repro steps

```C#
public record class MyType
{
    public bool Condition1 { get; set; }
    public bool Condition2 { get; set; }

    public int GetSomething()
    {
        return this switch
        {
            MyType
            {
                Condition1: true,
                Condition2: false
            } => 1,

            MyType
            {
                Condition2: true
            } => 2,

            _ => 0
        };
    }
}
```

### Related information

* .NET 6 / C# 10
* Visual Studio 2022
* SonarLint 6.15.0.67199

**Comment by martin-strecker-sonarsource** [comment]
Thank you for reporting this @prinz7. Can you please elaborate on why you added `MyType` in the pattern? It is optional and doesn't do anything in this case so the following code is equivalent (it compiles down to the same IL) and does not raise S3060:
```cs
public int GetSomething()
{
    return this switch
    {
        {
            Condition1: true,
            Condition2: false
        } => 1,
        {
            Condition2: true
        } => 2,

        _ => 0
    };
}
```

**Comment by prinz7** [comment]
Thank you for your answer.
I guess we used it this way for readability. And I didn't know that this change would remove the issue.
I guess it is a suitable solution.

Although, one may ask why the issue is raised in one case if the two cases are equivalent?
